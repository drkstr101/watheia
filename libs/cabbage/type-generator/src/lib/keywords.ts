import { factory, SyntaxKind } from 'typescript';

export const BOOLEAN = factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword);

export const DATE = factory.createTypeReferenceNode('Date');

export const NUMBER = factory.createKeywordTypeNode(SyntaxKind.NumberKeyword);

export const STRING = factory.createKeywordTypeNode(SyntaxKind.StringKeyword);

export const UNKNOWN = factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword);
