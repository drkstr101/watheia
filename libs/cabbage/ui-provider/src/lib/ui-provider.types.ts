export interface StackbitElement {
  'data-sb-field-path'?: string;
  'data-sb-object-id'?: string;
}
