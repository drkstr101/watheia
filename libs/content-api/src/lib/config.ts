const appUrl = process.env['APP_URL'] || 'https://watheia.vercel.com';
export default {
  apiUrl: process.env['API_URL'] || 'https://watheia.vercel.com/api',
  appUrl,
  snippetUrl: process.env['SNIPPET_URL'] || `${appUrl}/snippet.js`,
  controlUrl: process.env['CONTROL_URL'] || `${appUrl}/control.js`,
  logLevel: process.env['LOG_LEVEL'] || 'info',
};
