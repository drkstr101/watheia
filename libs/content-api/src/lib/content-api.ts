import type { ContentCache, LocalContentSchema } from './content-api.types';
import { resolveContent, resolveStaticPaths, resolveStaticProps } from './resolver';

export class LocalContentApi {
  public static async create(schema: LocalContentSchema): Promise<LocalContentApi> {
    return new LocalContentApi(await resolveContent(schema));
  }

  constructor(public readonly cache: ContentCache) {}

  resolvePaths(): string[] {
    return resolveStaticPaths(this.cache);
  }

  resolveProps(urlPath: string) {
    return resolveStaticProps(urlPath, this.cache);
  }
}
