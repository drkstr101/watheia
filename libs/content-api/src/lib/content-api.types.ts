import { DocumentEntry, Model, PageEntry, types } from '@watheia/content-model';

export interface ContentApi {
  resolve: () => ContentApi;
}

export interface LocalContentSchema {
  rootPath: string;
  models: Model[];
}

export interface MetaTag {
  property: string;
  content: string;
  format: 'property' | 'name';
}

type Context = {
  /** The document/asset file path relative to project directory. */
  filePath: string;
};

export type DocumentContext = Context;
export type AssetContext = Context;

export interface ContentCache {
  objects: DocumentEntry[];
  pages: PageEntry[];
  props: { site: types.Config };
}
