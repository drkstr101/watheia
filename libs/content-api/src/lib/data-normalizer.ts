// import { Model, ModelName } from '@watheia/content-model';
// import { LocalContentSchema } from '../content-api.types';

// type Entry = Record<string, unknown>;

// function getMetaData(data: Entry, model: Model, relPath: string) {
//   if (model.type === 'page') {
//     const urlPath = model.urlPath ?? '/{slug}';
//     const slug = (data['slug'] as string) ?? '';
//     return {
//       id: relPath,
//       type: model.type,
//       urlPath: urlPath.replace('{slug}', slug),
//     };
//   }

//   return { id: relPath, type: model.type };
// }

// export function withDataNormalizer(
//   schema: LocalContentSchema,
//   documentsById: Record<string, Entry>
// ) {
//   // collect reference fields for models
//   const referenceFields: Record<string, boolean> = {};
//   schema.models.forEach((model) => {
//     model.fields?.forEach((field) => {
//       if (
//         field.type === 'reference' ||
//         (field.type === 'list' && field.items?.type === 'reference')
//       ) {
//         referenceFields[model.name + ':' + field.name] = true;
//       }
//     });
//   });

//   const isRefField = (modelName: string, fieldName: string) =>
//     !!referenceFields[modelName + ':' + fieldName];

//   function normalizeData(data: Record<string, any>, model: Model, relPath: string) {
//     if (!data || !data['type']) return;

//     if (!data['__metadata']) data['__metadata'] = getMetaData(data, model, relPath);

//     const modelName = data['type'] as ModelName;
//     // Make Sourcebit-compatible
//     // if (!data.__metadata) data.__metadata = { modelName: data.type };

//     for (const fieldName in data) {
//       if (fieldName === '__metadata') continue;

//       let fieldValue = data[fieldName];
//       if (!fieldValue) continue;

//       const isRef = isRefField(modelName, fieldName);
//       if (Array.isArray(fieldValue)) {
//         if (fieldValue.length === 0) continue;
//         if (isRef && typeof fieldValue[0] === 'string') {
//           fieldValue = fieldValue.map((filename) => documentsById[filename]);
//           data[fieldName] = fieldValue;
//         }
//         if (typeof fieldValue[0] === 'object') {
//           fieldValue.forEach((o) => normalizeData(o, model, relPath));
//         }
//       } else {
//         if (isRef && typeof fieldValue === 'string') {
//           fieldValue = documentsById[fieldValue];
//           data[fieldName] = fieldValue;
//         }
//         if (typeof fieldValue === 'object') {
//           normalizeData(fieldValue, model, relPath);
//         }
//       }
//     }
//   }

//   return async ([relPath, data]: [string, Entry]) => {
//     // console.log(`Resolving:`, file);

//     const modelName = data['type'] as ModelName;
//     const model = schema.models.find((m) => m.name === modelName);
//     if (!model) {
//       throw new Error(`Unhandled model type (${modelName}) found in ${relPath}`);
//     }

//     normalizeData(data, model, relPath);
//   };
// }
