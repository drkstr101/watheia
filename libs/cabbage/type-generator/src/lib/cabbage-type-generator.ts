import {
  Field,
  FieldEnum,
  FieldList,
  FieldModel,
  FieldObject,
  FieldReference,
  Model,
} from '@watheia/content-model';
import sortBy from 'lodash/sortBy';
import {
  ListFormat,
  ScriptKind,
  ScriptTarget,
  SyntaxKind,
  TypeNode,
  createPrinter,
  createSourceFile,
  factory,
} from 'typescript';

import * as keywords from './keywords';

type FieldType = Field['type'];

const EXPORT_KEYWORD = factory.createModifier(SyntaxKind.ExportKeyword);

const OPTIONAL_TOKEN = factory.createToken(SyntaxKind.QuestionToken);

// const READ_ONLY_MODIFIER = factory.createModifier(SyntaxKind.ReadonlyKeyword);

function createUnion(...nodes: TypeNode[]) {
  return factory.createUnionTypeNode(nodes);
}

function createProperty(name: string, type: TypeNode, required = true) {
  return factory.createPropertySignature(
    undefined,
    name,
    required ? undefined : OPTIONAL_TOKEN,
    type
  );
}

function createLiteral(value: string | number) {
  const literal =
    typeof value === 'string'
      ? factory.createStringLiteral(value, true)
      : factory.createNumericLiteral(value);
  return factory.createLiteralTypeNode(literal);
}

function createMetadataProperty(model: Model) {
  const properties = [createProperty('modelType', createLiteral(model.type))];
  // append id prop for data/page types
  if (model.type === 'data' || model.type === 'page') {
    properties.push(createProperty('id', keywords.STRING));
  }

  // append urlPath prop for page types
  if (model.type === 'page') {
    properties.push(createProperty('urlPath', keywords.STRING));
  }

  return createProperty('__metadata', factory.createTypeLiteralNode(properties));
}

function createObjectType(field: FieldObject): TypeNode {
  return factory.createTypeLiteralNode(field.fields.map(fieldTransformer));
}

function createEnumType(field: FieldEnum): TypeNode {
  return factory.createUnionTypeNode(
    field.options.map((f) => {
      if (typeof f === 'string') {
        return createLiteral(f);
      } else if (typeof f === 'number') {
        return createLiteral(f);
      }

      return createLiteral(f.value);
    })
  );
}

function createModelType(field: FieldModel): TypeNode {
  return factory.createUnionTypeNode(
    field.models.map((name) => factory.createTypeReferenceNode(name))
  );
}

function createReferenceType(field: FieldReference): TypeNode {
  return factory.createUnionTypeNode(
    field.models.map((name) => factory.createTypeReferenceNode(name))
  );
}

function createListType(field: FieldList): TypeNode {
  // since we do not actually need a name to determine a field's type, we can
  // simulate one here to reuse our field type transformer code on the list items
  const items = { name: 'UNSPECIFIED', ...field.items };
  return factory.createArrayTypeNode(getFieldType(items));
}

function getFieldType(field: Field): TypeNode {
  switch (field.type) {
    case 'boolean':
      return keywords.BOOLEAN;
    case 'color':
      return createUnion(keywords.STRING, keywords.NUMBER);
    case 'cross-reference':
      return keywords.UNKNOWN;
    case 'date':
      return keywords.DATE;
    case 'datetime':
      return keywords.DATE;
    case 'enum':
      return createEnumType(field);
    case 'file':
      return keywords.UNKNOWN;
    case 'html':
      return keywords.STRING;
    case 'image':
      return keywords.STRING;
    case 'json':
      return keywords.UNKNOWN;
    case 'list':
      return createListType(field);
    case 'markdown':
      return keywords.STRING;
    case 'model':
      return createModelType(field);
    case 'number':
      return keywords.NUMBER;
    case 'object':
      return createObjectType(field);
    case 'reference':
      return createReferenceType(field);
    case 'richText':
      return keywords.STRING;
    case 'string':
      return keywords.STRING;
    case 'slug':
      return keywords.STRING;
    case 'style':
      // type: Record<string, Record<string, unkown>>
      return factory.createTypeReferenceNode('Record', [
        keywords.STRING,
        factory.createTypeReferenceNode('Record', [keywords.STRING, keywords.UNKNOWN]),
      ]);
    case 'text':
      return keywords.STRING;
    case 'url':
      return keywords.STRING;
    default:
      throw new Error(`Unsupported field detected: ${field}`);
  }
}

function modelTransformer(model: Model) {
  const __metadata = createMetadataProperty(model);
  const type = createProperty('type', createLiteral(model.name));
  const fields = (model.fields ?? []).map(fieldTransformer);
  return factory.createInterfaceDeclaration([EXPORT_KEYWORD], model.name, undefined, undefined, [
    __metadata,
    type,
    ...fields,
  ]);
}

function fieldTransformer(field: Field) {
  return createProperty(field.name, getFieldType(field), !!field.required);
}

export function generateModelTypes(models: Model[]) {
  const nodes = factory.createNodeArray(sortBy(models, (m) => m.name).map(modelTransformer));
  const sourceFile = createSourceFile(
    'placeholder.ts',
    '',
    ScriptTarget.ESNext,
    true,
    ScriptKind.TS
  );

  return createPrinter().printList(ListFormat.MultiLine, nodes, sourceFile);
}
