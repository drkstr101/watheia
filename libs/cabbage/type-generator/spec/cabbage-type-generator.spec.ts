import { generateModelTypes } from '../src/lib/cabbage-type-generator';
import model from './support/models';

describe('watheia.cabbage/type-generator', () => {
  // afterEach(() => rmSync(tempFile));
  describe('generateModelTypes', () => {
    it('generates an empty file when no models provided', () => {
      const output = generateModelTypes([]);
      expect(output).toBeEmpty();
    });

    it('transforms simple model types', () => {
      const output = generateModelTypes([
        {
          name: 'Page',
          type: 'page',
          fields: [
            { name: 'slug', type: 'string', required: true },
            { name: 'content', type: 'markdown' },
          ],
        },
      ]);
      expect(output).toEqualIgnoringWhitespace(`
        export interface Page {
          __metadata: { modelType: 'page'; id: string; urlPath: string; };
          type: 'Page';
          slug: string;
          content?: string;
        }
      `);
    });

    it('transforms complex model types', () => {
      const output = generateModelTypes(Object.values(model));
      expect(output).toEqualIgnoringWhitespace(`
        export interface Button {
            __metadata: {
                modelType: 'object';
            };
            type: 'Button';
            label?: string;
            url?: string;
            theme?:  'default' | 'outline';
        }
        export interface Hero {
            __metadata: {
                modelType: 'object';
            };
            type: 'Hero';
            heading?: string;
            body?: string;
            image?: {
              src?: string;
              alt?: string;
            };
            button?: Button;
            theme?:  'imgLeft' | 'imgRight';
        }
        export interface Page {
            __metadata: {
                modelType: 'page';
                id: string;
                urlPath: string;
            };
            type: 'Page';
            title: string;
            sections?: (Hero | Stats)[];
        }
        export interface Stats {
            __metadata: {
                modelType: 'object';
            };
            type: 'Stats';
            heading?: string;
            body?: string;
            stats?: { label?: string; value?: string; }[];
            theme?: 'primary' | 'dark';
        }
      `);
    });
  });
});
