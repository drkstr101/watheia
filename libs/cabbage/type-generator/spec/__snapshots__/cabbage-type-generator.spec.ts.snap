// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`watheia.cabbage/type-generator generateModelTypes transforms complex model types 1`] = `
"export interface Article {
    __metadata: {
        modelType: 'page';
        id: string;
        urlPath: string;
    };
    type: 'Article';
    title: string;
    slug: string;
    date: Date;
    author?: Person;
    excerpt?: string;
    featuredImage?: ImageBlock;
    content?: string;
    bottomSections?: (CarouselSection | DividerSection | FeaturedItemsSection | FeaturedPeopleSection | FeaturedPostsSection | GenericSection | ImageGallerySection | PricingSection | RecentPostsSection)[];
    isFeatured?: boolean;
    isDraft?: boolean;
    seo?: Seo;
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-neutralAlt-fg-dark' | 'bg-dark-fg-light';
    styles?: Record<string, Record<string, unknown>>;
}
export interface BackgroundImage {
    __metadata: {
        modelType: 'object';
    };
    type: 'BackgroundImage';
    url?: string;
    altText?: string;
    backgroundSize?: 'auto' | 'cover' | 'contain';
    backgroundPosition?: 'bottom' | 'center' | 'left' | 'left-bottom' | 'left-top' | 'right' | 'right-bottom' | 'right-top' | 'top';
    backgroundRepeat?: 'repeat' | 'repeat-x' | 'repeat-y' | 'no-repeat';
    opacity?: number;
}
export interface Badge {
    __metadata: {
        modelType: 'object';
    };
    type: 'Badge';
    label: string;
    color?: 'text-black' | 'text-white' | 'text-neutral' | 'text-primary';
    styles?: Record<string, Record<string, unknown>>;
}
export interface Button {
    __metadata: {
        modelType: 'object';
    };
    type: 'Button';
    label?: string;
    altText?: string;
    url: string;
    showIcon?: boolean;
    icon?: 'arrowDown' | 'arrowLeft' | 'arrowRight' | 'arrowUp' | 'chevronDown' | 'chevronLeft' | 'chevronBigLeft' | 'chevronRight' | 'chevronBigRight' | 'facebook' | 'github' | 'instagram' | 'linkedin' | 'mail' | 'play' | 'reddit' | 'send' | 'shoppingBag' | 'twitter' | 'vimeo' | 'youtube';
    iconPosition?: 'left' | 'right';
    style?: 'primary' | 'secondary';
    elementId?: string;
}
export interface CarouselSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'CarouselSection';
    title?: TitleBlock;
    subtitle?: string;
    items?: (FeaturedItem)[];
    badge?: Badge;
    elementId?: string;
    variant?: 'next-prev-nav' | 'dots-nav' | 'tabs-nav';
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface CheckboxFormControl {
    __metadata: {
        modelType: 'object';
    };
    type: 'CheckboxFormControl';
    name: string;
    label?: string;
    isRequired?: boolean;
    width?: 'full' | '1/2';
}
export interface Config {
    __metadata: {
        modelType: 'data';
        id: string;
    };
    type: 'Config';
    favicon?: string;
    header?: Header;
    footer?: Footer;
    titleSuffix?: string;
    defaultSocialImage?: string;
    defaultMetaTags?: (MetaTag)[];
}
export interface DividerSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'DividerSection';
    title?: string;
    elementId?: string;
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    styles?: Record<string, Record<string, unknown>>;
}
export interface EmailFormControl {
    __metadata: {
        modelType: 'object';
    };
    type: 'EmailFormControl';
    name?: string;
    label?: string;
    hideLabel?: boolean;
    placeholder?: string;
    isRequired?: boolean;
    width?: 'full' | '1/2';
}
export interface FeaturedItem {
    __metadata: {
        modelType: 'object';
    };
    type: 'FeaturedItem';
    title: string;
    tagline?: string;
    subtitle?: string;
    text?: string;
    image?: ImageBlock;
    actions?: (Button | Link)[];
    elementId?: string;
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-neutralAlt-fg-dark' | 'bg-dark-fg-light';
    styles?: Record<string, Record<string, unknown>>;
}
export interface FeaturedItemsSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'FeaturedItemsSection';
    title?: TitleBlock;
    subtitle?: string;
    items?: (FeaturedItem)[];
    actions?: (Button | Link)[];
    badge?: Badge;
    elementId?: string;
    variant?: 'two-col-grid' | 'three-col-grid' | 'small-list' | 'big-list' | 'toggle-list';
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface FeaturedPeopleSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'FeaturedPeopleSection';
    title?: TitleBlock;
    subtitle?: string;
    people?: (Person)[];
    actions?: (Button | Link)[];
    badge?: Badge;
    elementId?: string;
    variant?: 'three-col-grid' | 'four-col-grid' | 'mixed-grid';
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface FeaturedPostsSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'FeaturedPostsSection';
    title?: TitleBlock;
    subtitle?: string;
    posts?: (Article)[];
    showThumbnail?: boolean;
    showExcerpt?: boolean;
    showDate?: boolean;
    showAuthor?: boolean;
    actions?: (Button | Link)[];
    badge?: Badge;
    elementId?: string;
    variant?: 'two-col-grid' | 'three-col-grid' | 'small-list' | 'big-list';
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface Footer {
    __metadata: {
        modelType: 'data';
        id: string;
    };
    type: 'Footer';
    logo?: ImageBlock;
    title?: string;
    text?: string;
    primaryLinks?: FooterLinksGroup;
    secondaryLinks?: FooterLinksGroup;
    socialLinks?: (Social)[];
    legalLinks?: (Link)[];
    copyrightText?: string;
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    styles?: Record<string, Record<string, unknown>>;
}
export interface FooterLinksGroup {
    __metadata: {
        modelType: 'object';
    };
    type: 'FooterLinksGroup';
    title?: string;
    links?: (Button | Link)[];
    styles?: unknown;
}
export interface FormBlock {
    __metadata: {
        modelType: 'object';
    };
    type: 'FormBlock';
    fields?: (TextFormControl | EmailFormControl | TextareaFormControl | CheckboxFormControl | SelectFormControl)[];
    submitButton?: SubmitButtonFormControl;
    action?: string;
    destination?: string;
    elementId: string;
    styles?: Record<string, Record<string, unknown>>;
}
export interface GenericSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'GenericSection';
    title?: TitleBlock;
    subtitle?: string;
    text?: string;
    actions?: (Button | Link)[];
    media?: FormBlock | ImageBlock | VideoBlock;
    badge?: Badge;
    elementId?: string;
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface Header {
    __metadata: {
        modelType: 'data';
        id: string;
    };
    type: 'Header';
    title?: string;
    logo?: ImageBlock;
    primaryLinks?: (Button | Link | SubNav)[];
    secondaryLinks?: (Button | Link)[];
    variant?: 'logo-left-primary-nav-left' | 'logo-left-primary-nav-centered' | 'logo-left-primary-nav-right' | 'logo-centered-primary-nav-left' | 'logo-centered-primary-nav-centered';
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    styles?: Record<string, Record<string, unknown>>;
}
export interface ImageBlock {
    __metadata: {
        modelType: 'object';
    };
    type: 'ImageBlock';
    url?: string;
    altText?: string;
    elementId?: string;
    styles?: Record<string, Record<string, unknown>>;
}
export interface ImageGallerySection {
    __metadata: {
        modelType: 'object';
    };
    type: 'ImageGallerySection';
    title?: TitleBlock;
    subtitle?: string;
    images?: (ImageBlock)[];
    badge?: Badge;
    elementId?: string;
    motion?: 'static' | 'move-to-left' | 'move-to-right';
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface Link {
    __metadata: {
        modelType: 'object';
    };
    type: 'Link';
    label?: string;
    altText?: string;
    url: string;
    showIcon?: boolean;
    icon?: 'arrowDown' | 'arrowLeft' | 'arrowRight' | 'arrowUp' | 'chevronDown' | 'chevronLeft' | 'chevronBigLeft' | 'chevronRight' | 'chevronBigRight' | 'facebook' | 'github' | 'instagram' | 'linkedin' | 'mail' | 'play' | 'reddit' | 'send' | 'shoppingBag' | 'twitter' | 'vimeo' | 'youtube';
    iconPosition?: 'left' | 'right';
    style?: 'primary' | 'secondary';
    elementId?: string;
}
export interface MetaTag {
    __metadata: {
        modelType: 'object';
    };
    type: 'MetaTag';
    property?: 'og:title' | 'og:type' | 'og:image' | 'og:image:alt' | 'og:url' | 'og:description' | 'og:locale' | 'og:site_name' | 'og:video' | 'twitter:card' | 'twitter:site' | 'twitter:creator' | 'twitter:description' | 'twitter:title' | 'twitter:image' | 'twitter:image:alt' | 'twitter:player';
    content?: string;
}
export interface Page {
    __metadata: {
        modelType: 'page';
        id: string;
        urlPath: string;
    };
    type: 'Page';
    title: string;
    slug: string;
    sections?: (CarouselSection | DividerSection | FeaturedItemsSection | FeaturedPeopleSection | FeaturedPostsSection | GenericSection | ImageGallerySection | PricingSection | RecentPostsSection)[];
    isDraft?: boolean;
    seo?: Seo;
}
export interface PagedPostsSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'PagedPostsSection';
    title?: TitleBlock;
    subtitle?: string;
    showThumbnail?: boolean;
    showExcerpt?: boolean;
    showDate?: boolean;
    showAuthor?: boolean;
    actions?: (Button | Link)[];
    badge?: Badge;
    elementId?: string;
    variant?: 'two-col-grid' | 'three-col-grid' | 'small-list' | 'big-list';
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface Person {
    __metadata: {
        modelType: 'data';
        id: string;
    };
    type: 'Person';
    name: string;
    role?: string;
    bio?: string;
    image?: ImageBlock;
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-neutralAlt-fg-dark' | 'bg-dark-fg-light';
    elementId?: string;
    styles?: Record<string, Record<string, unknown>>;
}
export interface PostFeedLayout {
    __metadata: {
        modelType: 'page';
        id: string;
        urlPath: string;
    };
    type: 'PostFeedLayout';
    title?: string;
    slug: string;
    postFeed?: PagedPostsSection;
    numOfPostsPerPage?: number;
    enableSearch?: boolean;
    topSections?: (CarouselSection | DividerSection | FeaturedItemsSection | FeaturedPeopleSection | FeaturedPostsSection | GenericSection | ImageGallerySection | PricingSection | RecentPostsSection)[];
    bottomSections?: (CarouselSection | DividerSection | FeaturedItemsSection | FeaturedPeopleSection | FeaturedPostsSection | GenericSection | ImageGallerySection | PricingSection | RecentPostsSection)[];
    isDraft?: boolean;
    styles?: Record<string, Record<string, unknown>>;
    seo?: Seo;
}
export interface PostFeedSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'PostFeedSection';
    title?: TitleBlock;
    subtitle?: string;
    showThumbnail?: boolean;
    showExcerpt?: boolean;
    showDate?: boolean;
    showAuthor?: boolean;
    actions?: (Button | Link)[];
    badge?: Badge;
    elementId?: string;
    variant?: 'two-col-grid' | 'three-col-grid' | 'small-list' | 'big-list';
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface PricingPlan {
    __metadata: {
        modelType: 'object';
    };
    type: 'PricingPlan';
    title?: string;
    price?: string;
    details?: string;
    description?: string;
    features?: string[];
    image?: ImageBlock;
    actions?: (Button | Link)[];
    elementId?: string;
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-neutralAlt-fg-dark' | 'bg-dark-fg-light';
    styles?: Record<string, Record<string, unknown>>;
}
export interface PricingSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'PricingSection';
    title?: TitleBlock;
    subtitle?: string;
    plans?: (PricingPlan)[];
    badge?: Badge;
    elementId?: string;
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface RecentPostsSection {
    __metadata: {
        modelType: 'object';
    };
    type: 'RecentPostsSection';
    title?: TitleBlock;
    subtitle?: string;
    recentCount?: number;
    showThumbnail?: boolean;
    showExcerpt?: boolean;
    showDate?: boolean;
    showAuthor?: boolean;
    actions?: (Button | Link)[];
    badge?: Badge;
    elementId?: string;
    variant?: 'two-col-grid' | 'three-col-grid' | 'small-list' | 'big-list';
    colors?: 'bg-light-fg-dark' | 'bg-neutral-fg-dark' | 'bg-dark-fg-light';
    backgroundImage?: BackgroundImage;
    styles?: Record<string, Record<string, unknown>>;
}
export interface SelectFormControl {
    __metadata: {
        modelType: 'object';
    };
    type: 'SelectFormControl';
    name: string;
    label?: string;
    hideLabel?: boolean;
    defaultValue?: string;
    options?: string[];
    isRequired?: boolean;
    width?: 'full' | '1/2';
}
export interface Seo {
    __metadata: {
        modelType: 'object';
    };
    type: 'Seo';
    metaTitle?: string;
    metaDescription?: string;
    addTitleSuffix?: boolean;
    socialImage?: string;
    metaTags?: (MetaTag)[];
}
export interface Social {
    __metadata: {
        modelType: 'object';
    };
    type: 'Social';
    altText?: string;
    url: string;
    icon: 'facebook' | 'github' | 'instagram' | 'linkedin' | 'mail' | 'reddit' | 'twitter' | 'vimeo' | 'youtube';
    elementId?: string;
}
export interface SubNav {
    __metadata: {
        modelType: 'object';
    };
    type: 'SubNav';
    label?: string;
    altText?: string;
    links?: (Link)[];
    labelStyle?: 'primary' | 'secondary';
}
export interface SubmitButtonFormControl {
    __metadata: {
        modelType: 'object';
    };
    type: 'SubmitButtonFormControl';
    label?: string;
    showIcon?: boolean;
    icon?: 'arrowDown' | 'arrowLeft' | 'arrowRight' | 'arrowUp' | 'chevronDown' | 'chevronLeft' | 'chevronBigLeft' | 'chevronRight' | 'chevronBigRight' | 'facebook' | 'github' | 'instagram' | 'linkedin' | 'mail' | 'play' | 'reddit' | 'send' | 'shoppingBag' | 'twitter' | 'vimeo' | 'youtube';
    iconPosition?: 'left' | 'right';
    style?: 'primary' | 'secondary';
    elementId?: string;
}
export interface TextFormControl {
    __metadata: {
        modelType: 'object';
    };
    type: 'TextFormControl';
    name: string;
    label?: string;
    hideLabel?: boolean;
    placeholder?: string;
    isRequired?: boolean;
    width?: 'full' | '1/2';
}
export interface TextareaFormControl {
    __metadata: {
        modelType: 'object';
    };
    type: 'TextareaFormControl';
    name: string;
    label?: string;
    hideLabel?: boolean;
    placeholder?: string;
    isRequired?: boolean;
    width?: 'full' | '1/2';
}
export interface ThemeStyle {
    __metadata: {
        modelType: 'data';
        id: string;
    };
    type: 'ThemeStyle';
    light?: string | number;
    dark?: string | number;
    neutral?: string | number;
    neutralAlt?: string | number;
    primary?: string | number;
    fontBody?: 'sans' | 'serif';
    fontHeadlines?: 'sans' | 'serif';
    h1?: ThemeStyleHeading;
    h2?: ThemeStyleHeading;
    h3?: ThemeStyleHeading;
    h4?: ThemeStyleHeading;
    h5?: ThemeStyleHeading;
    h6?: ThemeStyleHeading;
    buttonPrimary?: ThemeStyleButton;
    buttonSecondary?: ThemeStyleButton;
    linkPrimary?: ThemeStyleLink;
    linkSecondary?: ThemeStyleLink;
}
export interface ThemeStyleButton {
    __metadata: {
        modelType: 'object';
    };
    type: 'ThemeStyleButton';
    weight?: 'normal' | 'medium' | 'bold';
    case?: 'none' | 'lowercase' | 'capitalize' | 'uppercase';
    letterSpacing?: 'tighter' | 'tight' | 'normal' | 'wide' | 'wider';
    borderRadius?: 'none' | 'DEFAULT' | 'lg' | 'xl' | 'full';
    shadow?: 'none' | 'md' | 'xl';
    horizontalPadding?: number;
    verticalPadding?: number;
}
export interface ThemeStyleHeading {
    __metadata: {
        modelType: 'object';
    };
    type: 'ThemeStyleHeading';
    size?: 'xs' | 'sm' | 'base' | 'lg' | 'xl' | '2xl' | '3xl' | '4xl' | '5xl' | '6xl' | '7xl' | '8xl' | '9xl';
    weight?: 'normal' | 'medium' | 'bold';
    decoration?: 'none' | 'underline' | 'line-through';
    case?: 'none' | 'lowercase' | 'capitalize' | 'uppercase';
    letterSpacing?: 'tighter' | 'tight' | 'normal' | 'wide' | 'wider';
}
export interface ThemeStyleLink {
    __metadata: {
        modelType: 'object';
    };
    type: 'ThemeStyleLink';
    weight?: 'normal' | 'medium' | 'bold';
    case?: 'none' | 'lowercase' | 'capitalize' | 'uppercase';
    letterSpacing?: 'tighter' | 'tight' | 'normal' | 'wide' | 'wider';
}
export interface TitleBlock {
    __metadata: {
        modelType: 'object';
    };
    type: 'TitleBlock';
    text?: string;
    color?: 'text-black' | 'text-white' | 'text-neutral' | 'text-primary';
    styles?: Record<string, Record<string, unknown>>;
}
export interface VideoBlock {
    __metadata: {
        modelType: 'object';
    };
    type: 'VideoBlock';
    title?: string;
    url?: string;
    autoplay?: boolean;
    loop?: boolean;
    muted?: boolean;
    controls?: boolean;
    aspectRatio?: '4:3' | '16:9';
    elementId?: string;
    styles?: Record<string, Record<string, unknown>>;
}
"
`;
